docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+```csharp
+public void EnsureCapacity<T>([NotNull]ref T[]? storage, int size)
+```
+
+The preceding code expresses the existing contract very clearly: Callers can pass a variable with the `null` value, but the return value is guaranteed to never be null. The `NotNull` attribute is most useful for `ref` and `out` arguments where `null` may be passed as an argument, but that argument is guaranteed to be not null when the method returns.
+
+## Specify conditional post-conditions: `NotNullWhen` and `MaybeNullWhen`
+
+You're likely familiar with the `string` method <xref:System.String.IsNullOrEmpty(string)?DisplayProperty=nameWithType>. This method returns `true` when the argument isn't null, and not the empty string. Callers shouldn't need to null-check the argument if the method returns `false`. To make a method like this nullable aware, you'd set the argument to a nullable type, and add the `NotNullWhen` attribute:
+
+```csharp
+bool IsNullOrEmpty([NotNullWhen(false)]string? value);
+```
+
+That informs the compiler that any code where the return value is `false` need not be checked:
+
+```csharp
+string? userInput = GetUserInput();
+if (!(string.IsNullOrEmpty(userInput))
+{
+   int messageLength = userInput.Length; // no null check needed.
+}
+// null check needed on userInput here.
+```
+
+The <xref:System.String.IsNullOrEmpty(string)?DisplayProperty=nameWithType> method will be annotated as shown above for .NET Core 3.0. You may have similar methods in your codebase that checks the state of objects for null values. The compiler won't recognize custom null check methods, and you'll need to add the annotations yourself.
+
+Another use for these attributes is the `Try*` pattern. The postconditions for `ref` and `out` variables are communicated through the return value. Consider this method shown earlier:
+
+```csharp
+bool TryGetMessage(string key, out string message)
+```
+
+The preceding method follows a typical .NET idiom: the return value indicates if `message` was set or not. If the method returns `true`, the value isn't null, `false`, and the method didn't set `message`.
+
+You can communicate that idiom using the `NotNullWhen` attribute. When you update the signature for nullable reference types, make `message` and `string?` and add an attribute:
+
+```csharp
+bool TryGetMessage(string key, [NotNullWhen(true)out string? message)
+```
+
+In the preceding example, the value of `message` is known to be not null when `TryGetMessage` returns true.  You should annotate similar methods in your codebase in the same way: the arguments could be `null`, and are known to be not null when the method returns `true`.
+
+There's one final attribute you may also need. Some methods manipulate string arguments. These methods will return a non-null string whenever the argument isn't null. To correctly annotate these methods, you use the `NotNullIfNotNull` attribute. Consider the following method:
```

Comment:
> Some methods manipulate string arguments.

I think strings are unrelated here. The core desire here is to have the output nullness be determined by the input nullness - if the input is nonnull, then the output should also be nonnull. This doesn't need to be a string, it could be any other data type where you want to have this kind of relationship

Reply:
Thanks. I mis-interpreted this badly. Updated.